scaper/core.py

```diff
@@ -1211,25 +1290,34 @@ def _instantiate_event(self, event, isbackground=False,
                             event_duration),
                         ScaperWarning)
 
-        # determine source time
-        source_time = -np.Inf
-        while source_time < 0:
-            source_time = _get_value_from_dist(event.source_time)
-
-        # Make sure source time + event duration is not greater than the
-        # source duration, if it is, adjust the source time (i.e. duration
-        # takes precedences over start time).
-        if source_time + event_duration > source_duration:
-            old_source_time = source_time
-            source_time = source_duration - event_duration
-            if not disable_instantiation_warnings:
+        # Modify event.source_time so that sampling from the source time distribution
+        # stays within the bounds of the audio file - event_duration. This allows users 
+        # to sample from anywhere in a source file without knowing the exact duration 
+        # of every source file. Only modify if label is not in protected labels.
+        if label not in self.protected_labels:
+            modified_source_time, warn = _ensure_satisfiable_source_time_tuple(
+                event.source_time, source_duration,  event_duration
+            )
+            
+            # determine source time and also check again just in case (for normal dist).
+            # if it happens again, just use the old method.
+            source_time = -np.Inf
+            while source_time < 0:
+                source_time = _get_value_from_dist(modified_source_time)
+                if source_time + event_duration > source_duration:
+                    source_time = source_duration - event_duration
+                    warn = True
+
+            if warn and not disable_instantiation_warnings:
                 warnings.warn(
-                    '{:s} source time ({:.2f}) is too great given event '
-                    'duration ({:.2f}) and source duration ({:.2f}), changed '
-                    'to {:.2f}.'.format(
-                        label, old_source_time, event_duration,
-                        source_duration, source_time),
+                    '{:s} source time tuple went out of bounds '
+                    'source duration ({:.2f}) and event duration ({:.2f}), modified source time '
+                    'tuple to stay within bounds'.format(
+                        label, source_duration, event_duration),
```

Comment:
Let's just be sure to update this warning to also print out the updated source_duration tuple

Reply:
See my comment on the previous thing about the warning.